### PR TITLE
feat(cli): generate system ABI's from World address

### PR DESCRIPTION
**Pull request by yonadaaa**
Adds a new MUD `generate-abi` command that generates an ABI for a given World by reading the onchain system registrations

fixes https://github.com/latticexyz/mud/issues/2845

### CODE REVIEW

File: packages/cli/src/commands/index.ts [hunk 1]
```diff
@@ -14,6 +14,7 @@ import test from "./test";
 import trace from "./trace";
 import devContracts from "./dev-contracts";
 import verify from "./verify";
+import abigen from "./abigen";
```
Comment (holic) [inline packages/cli/src/commands/index.ts:17]: not a huge fan of continuing the tablegen, worldgen, etc. naming pattern

I could see us eventually combining all of these features under a `codegen` command, maybe with sub commands like `codegen tables`), but in the meantime, something like "codegen abi" or "generate abi" for the CLI command and "world to abi" for the helper methods?

Comment (yonadaaa) [reply packages/cli/src/commands/index.ts:17]: hmm I like this line of thinking, lemme marinate on it

It's interesting that `abigen` and `worldgen` have different inputs but basically the same output

Comment (yonadaaa) [reply packages/cli/src/commands/index.ts]: The command is now `generate-abi` and the helper is just `systemToAbi`

File: packages/cli/src/commands/generate-abi.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  const worldAbi = [];
+
+  // render system ABI's
+  for (const system of systems) {
+    const fullOutputPath = `${resourceToLabel(system)}.json`;
+    const abi = systemToAbi(system);
+    await fs.writeFile(fullOutputPath, JSON.stringify(abi));
+
+    // prepare IWorld ABI
+    worldAbi.push(...abi);
```
Comment (holic) [inline packages/cli/src/commands/generate-abi.ts:58]: is this accurate?

systems will have a function name like `function increment()` but world may have it as `function app__increment()`

Comment (yonadaaa) [reply packages/cli/src/commands/generate-abi.ts:58]: wowww good point! i think we will have to prepare the world ABI seperately

File: packages/cli/src/commands/generate-abi.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    return yargs.options(generateAbiOptions);
+  },
+
+  async handler(args) {
+    await generateAbiHandler(args);
+    process.exit(0);
+  },
+};
+
+export async function generateAbiHandler(opts: Options) {
```
Comment (holic) [inline packages/cli/src/commands/generate-abi.ts:36]: thoughts on putting this into a separate dir like `packages/cli/src/init` or something to prepare for more of these types of files?

File: packages/cli/src/commands/generate-abi.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
+import type { CommandModule, InferredOptionTypes } from "yargs";
+import { Hex, createWalletClient, http } from "viem";
+import { getSystems } from "../deploy/getSystems";
+import { getWorldDeploy } from "../deploy/getWorldDeploy";
```
Comment (holic) [inline packages/cli/src/commands/generate-abi.ts:4]: we should consider "lifting" these out of `deploy` dir if we're sharing them across different commands/utils

Comment (yonadaaa) [reply packages/cli/src/commands/generate-abi.ts:4]: Yup, and I think having separate directories for each command makes less sense with these new tools. We'll basically be reusing everything except for the `ensureX` functions that actually deploy the contracts

File: packages/cli/src/commands/generate-abi.ts (hunk 3 above)
Comment (yonadaaa) [reply packages/cli/src/commands/generate-abi.ts:39]: What do you mean by "combining" the CLI commands? Like just putting some of this logic in a helper function? I imagine if we have one command that combines the logic of other commands, they all still live in the `commands` directory but share helpers

**Pull request closed without merging**